components/feature/sitepermissions/src/main/java/mozilla/components/feature/sitepermissions/SitePermissionsFeature.kt

```diff
@@ -96,32 +111,146 @@ class SitePermissionsFeature(
      * @param grantedPermissions the list of [grantedPermissions] that have been granted.
      */
     private fun onContentPermissionGranted(session: Session, grantedPermissions: List<Permission>) {
-        session.contentPermissionRequest.consume {
-            it.grant()
-            // Update the DB
+        session.contentPermissionRequest.consume { request ->
+            request.grant()
+            ioCoroutineScope.launch {
+                storeSitePermissions(request, grantedPermissions, ALLOWED)
+            }
             true
         }
     }
 
+    internal fun storeSitePermissions(
+        request: PermissionRequest,
+        permissions: List<Permission> = request.permissions,
+        status: SitePermissions.Status
+    ) {
+        var sitePermissions = storage.findSitePermissionsBy(request.host)
+
+        if (sitePermissions == null) {
+            sitePermissions = request.toSitePermissions(status = status, permissions = permissions)
+            storage.save(sitePermissions)
+        } else {
+            sitePermissions = request.toSitePermissions(status, sitePermissions)
+            storage.update(sitePermissions)
+        }
+    }
```

Comment:
It looks like this needs to be `synchronized` so that we are not updating the same permission concurrently?

Reply:
👍